### PR TITLE
Add exercise Circular Buffer

### PR DESCRIPTION
**Pull request by glennj**
I had a thought about how to implement this. It's a library like the list-ops exercise.

I'm curious what you all think.

Because it's a library, the tests don't use the bats `run` command, so `assert_success` and `assert_failure` don't apply. You'll see test assertions look like this
```bash
    buffer::write buff 'A'      && result=ok || result=fail
    assert_equal "$result" ok
```

I didn't write any additional instructions. It's a fairly difficult exercise, so users are on their own.

### CODE REVIEW

**Review by kotp — APPROVED** [review]

**Comment by kotp** [comment]
Get it out there and see what falls from the sky‽

**Comment by glennj** [comment]
I had a thought just now. What if, instead of 
```sh
buffer::overwrite bufname value
```
we had
```sh
buffer::write -f bufname value
```
That would get the student to to option handling inside the function.

That's a more shell-like syntax, but adds complexity. 

Worth it?

**Comment by kotp** [comment]
Is this the exercise we want to introduce additional complexity in?  It is already admittedly somewhat difficult.

File: exercises/practice/circular-buffer/.meta/example.sh [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+    buffer::incr_pointer "$name" read
+}
+
+buffer::write() {
+    local overwrite=false
+    # To use `getopts` in a function, these vars must be local.
+    local OPTIND OPTARG
+    while getopts :f opt; do
+        case $opt in
```
Comment (IsaacG) [inline exercises/practice/circular-buffer/.meta/example.sh:74]:
```suggestion
        case "$opt" in
```

**Review by IsaacG — APPROVED** [review]
The PR contents look good to me. The idea behind it...

I'm a bit iffy here based on the complexity. Just because you can get bash to do this doesn't mean it's reasonable to do so. I'm leaning more towards passing on exercises that don't fit the language vs jumping through hoops to get things working.

**Comment by glennj** [comment]
> I'm a bit iffy here based on the complexity. Just because you can get bash to do this doesn't mean it's reasonable to do so. I'm leaning more towards passing on exercises that don't fit the language vs jumping through hoops to get things working.

That is completely reasonable. I'm content to close this.

I'll create an issue so we can discuss foregoing exercises.

**Comment by glennj** [comment]
Although this is do-able in bash, it doesn't mean that it's smart to do it.

**Pull request closed without merging**